### ISSUE TITLE
Provision HTML from JSON (IResource) via middleware

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#3444]
Working issue part of #3202.

Basically, allowing provision of HTML in the default JSON-based `IResponseConverter` for HTML, via routing to installed, app-specific middleware, more or less corresponding to the pattern described [Multiple representations with middleware](https://github.com/per-samuelsson/Misc/wiki/Demystifying-request-handlers#multiple-representations-with-middleware).

**Comment by per-samuelsson** [comment]
Note: I'm separating this from the issue to later adapt client side Puppet to be more flexible, and to support merging, as discussed [here](https://github.com/Starcounter/Starcounter/issues/3407#issuecomment-196312062), in #3407.

**Comment by per-samuelsson** [comment]
Note to self: review state isolation of mime providers, to support a lock-free experience. Feasible to support registration of middleware only in the scope of the entrypoint (where requests to that app can not arrive)?

**Comment by per-samuelsson** [comment]
> Note to self: review state isolation of mime providers, to support a lock-free experience. Feasible to support registration of middleware only in the scope of the entrypoint (where requests to that app can not arrive)?

I instead implemented thread-safety by locking when providers are installed, and doing a copy in isolation, assigning that as the "main" copy when done, allowing us to keep reads lock-free.

**Comment by per-samuelsson** [comment]
Getting pretty much done with this one, although I can't merge/push it yet, because of #3445, not allowing me to do a personal build. If anyone feel a strong need to test it out before that, it's pushed to [issue/3444](https://github.com/Starcounter/Starcounter/tree/issue/3444) branch.

**Comment by warpech** [comment]
Looking forward to try it out!

**Comment by per-samuelsson** [comment]
> Looking forward to try it out!

I'll let you know when I can merge/push it, depending on #3445.

Also, we need to figure out #3446 to get to the next step. Everything else from Partial is now supported by middleware, except that.

**Comment by per-samuelsson** [comment]
Merged and pushed with https://github.com/Starcounter/Starcounter/commit/57c6d056c5a43e4ffabbea3f261e3f30219220b1.

The end result is pretty much as what is described in [Multiple representations with middleware](https://github.com/per-samuelsson/Misc/wiki/Demystifying-request-handlers#multiple-representations-with-middleware), and the chapter following that. The idea is to update that text to reflect the exact final result, but I'll wait with that until we've had the chance to play around with it, and get down to fix the set of related issues we need to do to fully be able to complement `Page` and ´Partial`, which include

1. Adapting HTML merger, getting rid of `IResource.GetHtmlPartialUrl()`, as outlined in #3407.
2. Make client-side puppet not rely blindly on that there is a "Html" property defined in the view-model.
3. Address #3446.

Until then, here's the short story enabled by the above merge:

```cs
void Main() {
  var htmlFromJson = new HtmlFromJsonProvider();
  Application.Current.Use(htmlFromJson);

  Handle.GET("/person", () => {
    return new Person();
  });
}
```

where the middleware can be tweaked in currently two ways:

```cs
class HtmlFromJsonProvider {
  /// <summary>
  /// Gets or sets a value if the current provider should provision "partial
  /// HTML" wrapped up in default standalone pages. Defaults to <c>false</c>.
  /// </summary>
  public bool ProvisionImplicitStandalonePages { get; set; }

  /// <summary>
  /// Gets or sets a value relaxing the provider to ignore any resource that
  /// does not expose a property referencing HTML. The default is <c>false</c>,
  /// and in such case, the provider will raise an error on any resource that
  /// miss a property referencing a HTML view path.
  /// </summary>
  public bool IgnoreJsonWithoutHtml { get; set; }
}
```

**Comment by per-samuelsson** [comment]
@warpech,

I'll merge this one into **pnext** too, and then await your feedback. I suggest you close this issue once you can use it to run "the simplest case", and it seem to work (even with rough edges). I suggest you make a branch of `https://github.com/warpech/SimplePuppet` that illustrate those changes, and play with it.

Then any bugs we find, or things you want to improve on, we keep in separate issues.

Until you tell me otherwise, I'll continue with something else after the merge to **pnext** is done. Let me know if you need any help!

**Comment by warpech** [comment]
Sure, I will try it out and get back to you!

On a sidenote, I dislike the fact that these 2 APIs differ, despite being so tightly related:

```
Application.Current.Use(htmlFromJson);
Handle.GET("/person", () => { .... });
```

State-of-the-art middleware-oriented web frameworks (KoaJS, ASP.NET Core ("vnext")) use middleware for anything, including HTTP routing. I guess we will get there eventually, but that is a matter for another set of issues.

**Comment by per-samuelsson** [comment]
> I'll merge this one into **pnext** too.

Tested and done.

**Comment by per-samuelsson** [comment]
> On a sidenote, **I dislike the fact that these 2 APIs differ** ...

Me too. The shift to build up the pipeline using middleware is just started. The idea is to take us the full way there in DNX/"vnext".

Also, in relation to that, we should all start to embrace and think more in terms of that what you do is in the context of an *application*. This new middleware supporting HTML from JSON provision is fully application-specific, while a lot of other stuff I've seen surely is not just yet.

**Comment by warpech** [comment]
I played with it according to your suggestion. The result is here: https://github.com/warpech/SimplePuppet/tree/middleware

**Comment by warpech** [comment]
I played with it according to your suggestion. The result is here: https://github.com/warpech/SimplePuppet/tree/middleware

My findings so far:

- Middleware is isolated between apps :+1:
- To use the implicit standalone mode (with implicit session), I have to set the ProvisionImplicitStandalonePages  property to `true`. 
- I can write my custom implicit standalone mode by creating a custom middleware class in my app. :+1:

This all is super cool! Just a few suggestions.

1. A big problem: I would like only some of my Jsons to be represented in Html and some remain just plain Json. This is illlustrated by https://github.com/warpech/SimplePuppet/blob/middleware/Program.cs#L16. Calling `/simple/raw` results in an exception: `ScErrInvalidOperation (SCERR1025): Operation invalid for the object's current state. Json instance Simple.Raw missing 'Html' property.` 
2. Would it be possible to split HtmlFromJsonProvider into 2 middlewares?
	- `HtmlFromJsonProvider`
	- `PuppetClientProvider` (that adds implicit standalone mode and implicit session -- which go of hand in hand)
3. Any idea on how to use middleware conditionally? Can I wrap middleware in another middleware that makes a conditional check?

**Comment by per-samuelsson** [comment]
Thanks for the feedback, @warpech. Just want to express that this initial thing I've provided is a *starter*, so this kind of feedback loop is what we now need to go through to get it into something we can later start to promote. There are certainly several things we can and should enhance further!

> A big problem: I would like only some of my Jsons to be represented in Html and some remain just plain Json.

Yeah, I was thinking about that too, when I implemented it. The design change to go from per-class to per-app change the scene a bit. I guess the way to go is to allow some kind of customization that allow a programmer to indicate what instances to exclude, but I'm not sure exactly how. Let's give it some thought.

As for

> ... this is illlustrated by https://github.com/warpech/SimplePuppet/blob/middleware/Program.cs#L16. Calling `/simple/raw` results in an exception: ScErrInvalidOperation (SCERR1025): Operation invalid for the object's current state. Json instance Simple.Raw missing 'Html' property.

You could instruct the middleware not to care about that a JSON don't come with a HTML property. See the tweaking section at the bottom of this https://github.com/Starcounter/Starcounter/issues/3444#issuecomment-197301317. But I guess that's not enough as an exclusion policy, right?

> Would it be possible to split HtmlFromJsonProvider into 2 middlewares?

I think so, yes. I need to give it some consideration though. Would it be a nicer experience you think?

> Any idea on how to use middleware conditionally?

Like I wrote before, middleware in the form of middleware classes (i.e. excluding request- and response-filters), is merely a pattern. Using that pattern, I have added middleware that support general MIME type providers, and built `HtmlFromJsonProvider` on top of that. Can you describe the exact case you are thinking about?

**Comment by warpech** [comment]
> Can you describe the exact case you are thinking about?

I was thinking to make a conditional middleware to only use  `HtmlFromJsonProvider` if the url equals `/simple`.

**Comment by per-samuelsson** [comment]
> I was thinking to make a conditional middleware to only use  `HtmlFromJsonProvider` if the url equals `/simple`.

I see. That would require three changes in the design:

1. Mime providers must get access not only to the resource, but also to the request.
2. We must support multiple providers for a single MIME type.
3. We must support chaining of mime providers, so that they can invoke (or choose not to invoke) the next provider in the chain.

As for [1], that's a pretty simple redesign. But [2] and [3] is more work, but certainly doable.

**Comment by per-samuelsson** [comment]
Continuing from where I left off in the above comment, I think we should materialize [1] no matter what. It just make sense, and I think it's fair that any MIME provider get the chance to consider the request.

An alternative to [2]-[3] could be to just introduce a filtering capability to `HtmlFromJsonProvider` isolated, and not touch underlying more general MIME providers. I'm thinking something like this:

```cs
void Main() {
  var viewModelViewProvider = new HtmlFromJsonProvider();
  viewModelViewProvider.ShouldInvoke((request, resource) => {
    if (request.RelativeUri == "/simple") {
      return true;
    }
    return false;
  }
  Application.Current.Use(viewModelViewProvider);
}
```

The benefit of that is that it's a simpler change. The drawback is that it's not general and we can use it to nothing else than this particular case.

**Comment by per-samuelsson** [comment]
Expanding yet further on the above two comments, if we decide to do the more general [2]-[3] (allowing multiple, chained providers / MIME type), we could make that in a way that also enables this:

> Would it be possible to split HtmlFromJsonProvider into 2 middlewares?

The experience that would come from that would be something like this:

```cs
void Main() {
  var viewModelViewProvider = new HtmlFromJsonProvider();
  var puppetViewProvider = new PuppetViewProvider();
  var app = Application.Current;
  app.Use(viewModelViewProvider);
  app.Use(puppetViewProvider);
}
```

And the former would simply invoke the later.

That would mean another extension to providers though: that in a chain, later providers must be given access to the result of former providers (which kind of make sense).

**Comment by warpech** [comment]
> And the former would simply invoke the later.

I was thinking exactly about this. Basically, `puppetViewProvider` would only work if declared in this order:

```
app.Use(viewModelViewProvider);
app.Use(puppetViewProvider);
```

But NOT in this order:

```
app.Use(puppetViewProvider);
app.Use(viewModelViewProvider);
```

This seems to be accepted in [ExpressJS][1] and in [general][2]. In [Django][3] it is even directly recommended.

BTW I have found some interesting articles about middleware patterns: https://github.com/ring-clojure/ring/wiki/Middleware-Patterns, http://www.alexedwards.net/blog/making-and-using-middleware

[1]: https://www.quora.com/Is-it-a-bad-practice-to-write-express-middleware-that-depends-on-other-express-middleware
[2]: http://programmers.stackexchange.com/questions/285128/middleware-dependencies
[3]: https://docs.djangoproject.com/en/1.9/topics/http/middleware/

**Comment by per-samuelsson** [comment]
Thanks for sharing, @warpech.

I'm interpreting your responses as you think we should go with the more general approach, not the shortcut I suggested to introduce filtering in `HtmlFromJsonProvider` in isolation.

I'm leaning towards that too. Especially so in the light of that we want to go the full way towards how middleware is designed in DNX, and the suggested pattern here is very much in line with that.

We still don't reach full compatibility, since the only concrete general middleware we have here are the mime providers I've introduced, and they are kind of boxed into the quite complex request pipeline of the server as it works today. But every step towards a general feeling of middleware like we want the end-result to appear seem wise to take if you ask me.

So let's decide here and now that we extend our middleware initiative to include what I suggested in this https://github.com/Starcounter/Starcounter/issues/3444#issuecomment-197763316 from earlier this morning, plus this addition:

> That would mean another extension to providers though: that in a chain, later providers must be given access to the result of former providers (which kind of make sense).

Four points in total. Five if we want to mix in the actual separation of `HtmlFromJsonProvider` into two separate middlewares, one for puppet provision. I'll wrap those up in a separate issue and get going on that right now.

With that, I think we can close this issue, right?

**Comment by per-samuelsson** [comment]
Oh, a thing I forgot to comment.

> var puppetViewProvider = new PuppetViewProvider();

@warpech, are you guys totally sure we are sticking to that Puppet terminology (the one I've been fighting against for so long :sunglasses:). Because if we go that way (in #3451), we are introducing that term basically on the first few lines of code in every sample- and prefab app we've got, and it's certainly not a term users will feel familiar with (or can in anyway take a guess on what it does), so we have point them to the documentation right away for that.

Is `PartialViewProvider` a better alternative? And on the subject, is `HtmlFromJsonProvider` a bad name? I also considered `ViewFromViewModelProvider`. What I like about the former is that include name of MIME types (in the context of mime providers).

**Comment by warpech** [comment]
> @warpech, are you guys totally sure we are sticking to that Puppet terminology (the one I've been fighting against for so long :sunglasses:).

Good catch.... I totally would like to avoid making this decision by myself. I urge to discuss this with @Starcounter-Jack and @chrhol.

Just keep in mind that we are speaking here about giving the name to the middleware that injects the proper **client side**, not about giving a name to the server side.

- The server side implementation of this pattern is nameless in code (though commonly referred as "Puppeteer").
- The client side implementation of this patter is already called "PuppetJs" everywhere in our docs. It is implemented in a JavaScript library called PuppetJs wrapped in a Web Component called `<puppet-client>`.

With that said, let me recap what `ViewPuppetProvider` does:

1. creates a `Session` (if not created before) that has a long life span and can be modified using JSON-Patch
2. wraps HTML partial (if not wrapped before) in a minimum HTML document, that creates an instance of PuppetJS
3. in client side, PuppetJS creates the WebSocket connection and maintains JSON-Patch synchronization between the client and the server.

This set of technologies (1-2-3) is what we call "Puppet". Otherwise the correct name is `ImplicitSessionWithHTMLWrappingOfThePartialsThatEnablesJsonSynchronizationProvider`.

Maybe we could put the `Session` part into a separate middleware, so then it becomes:

```
app.Use(viewModelViewProvider);
app.Use(viewModelSessionProvider);
app.Use(viewPuppetProvider);
```

> And on the subject, is HtmlFromJsonProvider a bad name? I also considered ViewFromViewModelProvider. What I like about the former is that include name of MIME types (in the context of mime providers).

I prefer `HtmlFromJsonProvider`, since the class it is using is called `Json`, not `ViewModel`.

**Comment by per-samuelsson** [comment]
>  I totally would like to avoid making this decision by myself. 

I fully understand that ...

> I urge to discuss this with @Starcounter-Jack and @chrhol.

... but I think I won't, because I've said it so many times over the years already, so I'm just gonna rest my case.

> Just keep in mind that we are speaking here about giving the name to the middleware that injects the proper **client side**, not about giving a name to the server side.

Well, if we name the middleware like it is proposed, every **server side** application will for sure expose it to anyone taking a look at the code. Which means that if you take a glimpse at any of our samples, coming from outside, you'll see this pattern:

```cs
void Main() {
  var app = Application.Current;
  app.Use(new HtmlFromJsonProvider());
  app.Use(new PuppetViewProvider());

  Handle.GET("/products", () => { ... });
}
```

And this is what I've always objected against, since I can read any code and **take a qualified guess** what it does just **by it's name** (in the context of web applications). But not for "Puppet". I would have to go read about it to understand it's role. 

If the name just gave *some* better hint, I might not need that, i.e:

```cs
void Main() {
  var app = Application.Current;
  app.Use(new HtmlFromJsonProvider());
  app.Use(new ClientServerSynchronizer());
  // or: app.Use(new JsonPatchSynchronization());
  // or: app.Use(new ViewModelPatch());

  Handle.GET("/products", () => { ... });
}
```

Well, you get the point. :smile: 

> The client side implementation of this pattern is already called "PuppetJs" everywhere in our docs. It is implemented in a JavaScript library called PuppetJs wrapped in a Web Component called <puppet-client>.

Well, yeah, sure, but isn't it like what they use to say about lies: just because you keep repeating them don't make them any more true. :wink:

**Comment by per-samuelsson** [comment]
> I prefer `HtmlFromJsonProvider`, since the class it is using is called `Json`, not `ViewModel`.

Good, I agree. Lets stick to that.

**Comment by per-samuelsson** [comment]
This whole thing is now done. Including the aftermath, isolated in #3446 and #3451.

**Comment by per-samuelsson** [comment]
Now I've also updated and extended the original article draft (chapter, [Multiple representations with middleware](https://github.com/per-samuelsson/Misc/wiki/Demystifying-request-handlers#multiple-representations-with-middleware)) to support our new extensions, and also to contain a new section on a few important [implementation details](https://github.com/per-samuelsson/Misc/wiki/Demystifying-request-handlers#implementation-details).
